cannot build the Types: got type "map[string]string" for field "Data", only types "string" and "*string" supported as sensitive

<!--
Thank you for helping to improve Terrajet!

Please be sure to search for open issues before raising a new one. We use issues
for bug reports and feature requests. Please find us at https://slack.crossplane.io
for questions, support, and discussion.
-->

### What happened?

The attempt to generate `rancher2_secret` resource in the context of https://github.com/AbsaOSS/provider-jet-rancher failed with
```
make generate
16:39:29 [ .. ] verify dependencies have expected content
all modules verified
16:39:34 [ OK ] go modules dependencies verified
16:39:34 [ .. ] go generate darwin_amd64
panic: cannot generate crd for resource rancher2_secret: cannot build types for Secret: cannot build the Types: got type "map[string]string" for field "Data", only types "string" and "*string" supported as sensitive

goroutine 1 [running]:
github.com/crossplane/terrajet/pkg/pipeline.Run(0xc000160600, 0xc000832570, 0x28)
	/Users/xxx/absa/provider-jet-rancher/.work/pkg/pkg/mod/github.com/turkenh/terrajet@v0.0.0-20220114225802-9f11209e5cac/pkg/pipeline/run.go:76 +0x1db3
main.main()
	/Users/xxx/absa/provider-jet-rancher/cmd/generator/main.go:48 +0xad
exit status 2
apis/generate.go:32: running "go": exit status 1
```


### How can we reproduce it?

Clone  https://github.com/AbsaOSS/provider-jet-rancher, remove limiting `WithIncludeList` from `config/provider.go` and run `make generate`

--- a/go.mod
+++ b/go.mod
@@ -11,7 +11,7 @@ require (
 	github.com/hashicorp/terraform-plugin-sdk/v2 v2.7.0
 	github.com/iancoleman/strcase v0.2.0
 	github.com/json-iterator/go v1.1.11
-	github.com/muvaf/typewriter v0.0.0-20210910160850-80e49fe1eb32
+	github.com/muvaf/typewriter v0.0.0-20220131201631-921e94e8e8d7
 	github.com/pkg/errors v0.9.1
 	github.com/spf13/afero v1.6.0
 	golang.org/x/tools v0.1.5

--- a/pkg/resource/sensitive.go
+++ b/pkg/resource/sensitive.go
@@ -33,11 +33,13 @@ import (
 )
 
 const (
-	errCannotExpandWildcards          = "cannot expand wildcards"
-	errFmtCannotGetValueForFieldPath  = "cannot not get a value for fieldpath %q"
-	errFmtCannotGetStringForFieldPath = "cannot not get a string for fieldpath %q"
-	errFmtCannotGetSecretKeySelector  = "cannot get SecretKeySelector from xp resource for fieldpath %q"
-	errFmtCannotGetSecretValue        = "cannot get secret value for %v"
+	errCannotExpandWildcards               = "cannot expand wildcards"
+	errFmtCannotGetValueForFieldPath       = "cannot not get a value for fieldpath %q"
+	errFmtCannotGetStringForFieldPath      = "cannot not get a string for fieldpath %q"
+	errFmtCannotGetSecretKeySelector       = "cannot get SecretKeySelector from xp resource for fieldpath %q"
+	errFmtCannotGetSecretKeySelectorAsList = "cannot get SecretKeySelector list from xp resource for fieldpath %q"
+	errFmtCannotGetSecretKeySelectorAsMap  = "cannot get SecretKeySelector map from xp resource for fieldpath %q"
+	errFmtCannotGetSecretValue             = "cannot get secret value for %v"
 )
 
 const (
@@ -204,7 +206,7 @@ func GetSensitiveParameters(ctx context.Context, client SecretClient, from runti
 			case map[string]interface{}:
 				sel := &map[string]v1.SecretKeySelector{}
 				if err = pavedJSON.GetValueInto(expandedJSONPath, sel); err != nil {
-					return errors.Wrapf(err, errFmtCannotGetSecretKeySelector, expandedJSONPath)
+					return errors.Wrapf(err, errFmtCannotGetSecretKeySelectorAsMap, expandedJSONPath)
 				}
 				sensitives := make(map[string]interface{})
 				for key, value := range *sel {
@@ -220,7 +222,7 @@ func GetSensitiveParameters(ctx context.Context, client SecretClient, from runti
 			case []interface{}:
 				sel := &[]v1.SecretKeySelector{}
 				if err = pavedJSON.GetValueInto(expandedJSONPath, sel); err != nil {
-					return errors.Wrapf(err, errFmtCannotGetSecretKeySelector, expandedJSONPath)
+					return errors.Wrapf(err, errFmtCannotGetSecretKeySelectorAsList, expandedJSONPath)
 				}
 				var sensitives []interface{}
 				for _, s := range *sel {

--- a/pkg/resource/sensitive.go
+++ b/pkg/resource/sensitive.go
@@ -142,14 +142,17 @@ func GetSensitiveAttributes(from map[string]interface{}, mapping map[string]stri
 				vals = map[string][]byte{}
 			}
 			switch s := v.(type) {
+			case map[string]interface{}:
+				for i, e := range s {
+					if err := setSensitiveAttributesToValuesMap(e, i, k, fp, vals); err != nil {
+						return nil, err
+					}
+				}
 			case []interface{}:
 				for i, e := range s {
-					value, ok := e.(string)
-					if !ok {
-						return nil, errors.Errorf(errFmtCannotGetStringForFieldPath, fp)
+					if err := setSensitiveAttributesToValuesMap(e, i, k, fp, vals); err != nil {
+						return nil, err
 					}
-					k = strings.TrimSuffix(k, pluralSuffix)
-					vals[fmt.Sprintf("%s%s.%v", prefixAttribute, k, i)] = []byte(value)
 				}
 			case string:
 				vals[fmt.Sprintf("%s%s", prefixAttribute, k)] = []byte(s)
@@ -198,7 +201,23 @@ func GetSensitiveParameters(ctx context.Context, client SecretClient, from runti
 			}
 
 			switch v.(type) {
-			case []map[string]interface{}, []interface{}:
+			case map[string]interface{}:
+				sel := &map[string]v1.SecretKeySelector{}
+				if err = pavedJSON.GetValueInto(expandedJSONPath, sel); err != nil {
+					return errors.Wrapf(err, errFmtCannotGetSecretKeySelector, expandedJSONPath)
+				}
+				sensitives := make(map[string]interface{})
+				for key, value := range *sel {
+					sensitive, err = client.GetSecretValue(ctx, value)
+					if err != nil {
+						return errors.Wrapf(err, errFmtCannotGetSecretValue, sel)
+					}
+					sensitives[key] = string(sensitive)
+				}
+				if err := setSensitiveParametersWithPaved(pavedTF, expandedJSONPath, tfPath, mapping, sensitives); err != nil {
+					return err
+				}
+			case []interface{}:
 				sel := &[]v1.SecretKeySelector{}
 				if err = pavedJSON.GetValueInto(expandedJSONPath, sel); err != nil {
 					return errors.Wrapf(err, errFmtCannotGetSecretKeySelector, expandedJSONPath)
@@ -214,7 +233,7 @@ func GetSensitiveParameters(ctx context.Context, client SecretClient, from runti
 				if err := setSensitiveParametersWithPaved(pavedTF, expandedJSONPath, tfPath, mapping, sensitives); err != nil {
 					return err
 				}
-			case map[string]interface{}, interface{}:
+			case interface{}:
 				sel := &v1.SecretKeySelector{}
 				if err = pavedJSON.GetValueInto(expandedJSONPath, sel); err != nil {
 					return errors.Wrapf(err, errFmtCannotGetSecretKeySelector, expandedJSONPath)
@@ -369,5 +388,15 @@ func setSensitiveParametersWithPaved(pavedTF *fieldpath.Paved, expandedJSONPath,
 	if err = pavedTF.SetValue(expTF, sensitives); err != nil {
 		return errors.Wrapf(err, "cannot set string as terraform attribute for fieldpath %q", tfPath)
 	}
+	return nil
+}
+
+func setSensitiveAttributesToValuesMap(e, i interface{}, k, fp string, vals map[string][]byte) error {
+	k = strings.TrimSuffix(k, pluralSuffix)
+	value, ok := e.(string)
+	if !ok {
+		return errors.Errorf(errFmtCannotGetStringForFieldPath, fp)
+	}
+	vals[fmt.Sprintf("%s%s.%v", prefixAttribute, k, i)] = []byte(value)
 	return nil
 }

--- a/pkg/resource/sensitive_test.go
+++ b/pkg/resource/sensitive_test.go
@@ -81,6 +81,28 @@ var (
 	errBoom = errors.New("boom")
 )
 
+type secretKeySelectorModifier func(s *xpv1.SecretKeySelector)
+
+func secretKeySelectorWithKey(v string) secretKeySelectorModifier {
+	return func(s *xpv1.SecretKeySelector) {
+		s.Key = v
+	}
+}
+
+func secretKeySelectorWithSecretReference(v xpv1.SecretReference) secretKeySelectorModifier {
+	return func(s *xpv1.SecretKeySelector) {
+		s.SecretReference = v
+	}
+}
+
+func secretKeySelector(sm ...secretKeySelectorModifier) *xpv1.SecretKeySelector {
+	s := &xpv1.SecretKeySelector{}
+	for _, m := range sm {
+		m(s)
+	}
+	return s
+}
+
 func TestGetConnectionDetails(t *testing.T) {
 	type args struct {
 		tr   Terraformed
@@ -144,6 +166,32 @@ func TestGetConnectionDetails(t *testing.T) {
 					"attribute.top_level_secret.0": []byte("val1"),
 					"attribute.top_level_secret.1": []byte("val2"),
 					"attribute.top_level_secret.2": []byte("val3"),
+				},
+			},
+		},
+		"Map": {
+			args: args{
+				tr: &fake.Terraformed{
+					MetadataProvider: fake.MetadataProvider{
+						ConnectionDetailsMapping: map[string]string{
+							"top_level_secrets": "status.atProvider.topLevelSecrets",
+						},
+					},
+				},
+				cfg: config.DefaultResource("terrajet_resource", nil),
+				data: map[string]interface{}{
+					"top_level_secrets": map[string]interface{}{
+						"key1": "val1",
+						"key2": "val2",
+						"key3": "val3",
+					},
+				},
+			},
+			want: want{
+				out: map[string][]byte{
+					"attribute.top_level_secret.key1": []byte("val1"),
+					"attribute.top_level_secret.key2": []byte("val2"),
+					"attribute.top_level_secret.key3": []byte("val3"),
 				},
 			},
 		},
@@ -346,11 +394,11 @@ func TestGetSensitiveAttributes(t *testing.T) {
 		},
 		"NotAValue": {
 			args: args{
-				paths: map[string]string{"top_config_secretmap": ""},
+				paths: map[string]string{"inner_optional": ""},
 				data:  testInput,
 			},
 			want: want{
-				err: errors.Errorf(errFmtCannotGetStringForFieldPath, "top_config_secretmap"),
+				out: nil,
 			},
 		},
 		"UnexpectedWildcard": {
@@ -415,11 +463,13 @@ func TestGetSensitiveParameters(t *testing.T) {
 					Object: map[string]interface{}{
 						"spec": map[string]interface{}{
 							"forProvider": map[string]interface{}{
-								"adminPasswordSecretRef": map[string]interface{}{
-									"name":      "admin-password",
-									"namespace": "crossplane-system",
-									"key":       "pass",
-								},
+								"adminPasswordSecretRef": secretKeySelector(
+									secretKeySelectorWithKey("pass"),
+									secretKeySelectorWithSecretReference(xpv1.SecretReference{
+										Name:      "admin-password",
+										Namespace: "crossplane-system",
+									}),
+								),
 							},
 						},
 					},
@@ -460,17 +510,21 @@ func TestGetSensitiveParameters(t *testing.T) {
 					Object: map[string]interface{}{
 						"spec": map[string]interface{}{
 							"forProvider": map[string]interface{}{
-								"passwordsSecretRef": []map[string]interface{}{
-									{
-										"name":      "db-passwords",
-										"namespace": "crossplane-system",
-										"key":       "admin",
-									},
-									{
-										"name":      "db-passwords",
-										"namespace": "crossplane-system",
-										"key":       "system",
-									},
+								"passwordsSecretRef": []interface{}{
+									secretKeySelector(
+										secretKeySelectorWithKey("admin"),
+										secretKeySelectorWithSecretReference(xpv1.SecretReference{
+											Name:      "db-passwords",
+											Namespace: "crossplane-system",
+										}),
+									),
+									secretKeySelector(
+										secretKeySelectorWithKey("system"),
+										secretKeySelectorWithSecretReference(xpv1.SecretReference{
+											Name:      "db-passwords",
+											Namespace: "crossplane-system",
+										}),
+									),
 								},
 							},
 						},
@@ -493,6 +547,65 @@ func TestGetSensitiveParameters(t *testing.T) {
 				},
 			},
 		},
+		"SingleNoWildcardWithMap": {
+			args: args{
+				clientFn: func(client *mocks.MockSecretClient) {
+					client.EXPECT().GetSecretValue(gomock.Any(), gomock.Eq(xpv1.SecretKeySelector{
+						SecretReference: xpv1.SecretReference{
+							Name:      "db-passwords",
+							Namespace: "crossplane-system",
+						},
+						Key: "admin",
+					})).Return([]byte("admin_pwd"), nil)
+					client.EXPECT().GetSecretValue(gomock.Any(), gomock.Eq(xpv1.SecretKeySelector{
+						SecretReference: xpv1.SecretReference{
+							Name:      "db-passwords",
+							Namespace: "crossplane-system",
+						},
+						Key: "system",
+					})).Return([]byte("system_pwd"), nil)
+				},
+				from: &unstructured.Unstructured{
+					Object: map[string]interface{}{
+						"spec": map[string]interface{}{
+							"forProvider": map[string]interface{}{
+								"passwordsSecretRef": map[string]interface{}{
+									"pwd1": secretKeySelector(
+										secretKeySelectorWithKey("admin"),
+										secretKeySelectorWithSecretReference(xpv1.SecretReference{
+											Name:      "db-passwords",
+											Namespace: "crossplane-system",
+										}),
+									),
+									"pwd2": secretKeySelector(
+										secretKeySelectorWithKey("system"),
+										secretKeySelectorWithSecretReference(xpv1.SecretReference{
+											Name:      "db-passwords",
+											Namespace: "crossplane-system",
+										}),
+									),
+								},
+							},
+						},
+					},
+				},
+				into: map[string]interface{}{
+					"some_other_key": "some_other_value",
+				},
+				mapping: map[string]string{
+					"db_passwords": "spec.forProvider.passwordsSecretRef",
+				},
+			},
+			want: want{
+				out: map[string]interface{}{
+					"some_other_key": "some_other_value",
+					"db_passwords": map[string]interface{}{
+						"pwd1": "admin_pwd",
+						"pwd2": "system_pwd",
+					},
+				},
+			},
+		},
 		"MultipleNoWildcard": {
 			args: args{
 				clientFn: func(client *mocks.MockSecretClient) {
@@ -508,11 +621,13 @@ func TestGetSensitiveParameters(t *testing.T) {
 					Object: map[string]interface{}{
 						"spec": map[string]interface{}{
 							"forProvider": map[string]interface{}{
-								"adminPasswordSecretRef": map[string]interface{}{
-									"name":      "admin-password",
-									"namespace": "crossplane-system",
-									"key":       "pass",
-								},
+								"adminPasswordSecretRef": secretKeySelector(
+									secretKeySelectorWithKey("pass"),
+									secretKeySelectorWithSecretReference(xpv1.SecretReference{
+										Name:      "admin-password",
+										Namespace: "crossplane-system",
+									}),
+								),
 							},
 						},
 					},
@@ -557,11 +672,13 @@ func TestGetSensitiveParameters(t *testing.T) {
 								"databaseUsers": []interface{}{
 									map[string]interface{}{
 										"name": "admin",
-										"passwordSecretRef": map[string]interface{}{
-											"name":      "admin-password",
-											"namespace": "crossplane-system",
-											"key":       "pass",
-										},
+										"passwordSecretRef": secretKeySelector(
+											secretKeySelectorWithKey("pass"),
+											secretKeySelectorWithSecretReference(xpv1.SecretReference{
+												Name:      "admin-password",
+												Namespace: "crossplane-system",
+											}),
+										),
 										"displayName": "Administrator",
 									},
 									map[string]interface{}{
@@ -578,11 +695,13 @@ func TestGetSensitiveParameters(t *testing.T) {
 									},
 									map[string]interface{}{
 										"name": "maintenance",
-										"passwordSecretRef": map[string]interface{}{
-											"name":      "maintenance-password",
-											"namespace": "crossplane-system",
-											"key":       "pass",
-										},
+										"passwordSecretRef": secretKeySelector(
+											secretKeySelectorWithKey("pass"),
+											secretKeySelectorWithSecretReference(xpv1.SecretReference{
+												Name:      "maintenance-password",
+												Namespace: "crossplane-system",
+											}),
+										),
 										"displayName": "Maintenance",
 									},
 								},

--- a/pkg/types/builder.go
+++ b/pkg/types/builder.go
@@ -156,8 +156,8 @@ func (g *Builder) buildResource(res *schema.Resource, cfg *config.Resource, tfPa
 			cfg.Sensitive.AddFieldPath(fieldPathWithWildcard(tfPaths), "spec.forProvider."+fieldPathWithWildcard(xpPaths)+sfx)
 			// todo(turkenh): do we need to support other field types as sensitive?
 			if fieldType.String() != "string" && fieldType.String() != "*string" && fieldType.String() != "[]string" &&
-				fieldType.String() != "[]*string" {
-				return nil, nil, fmt.Errorf(`got type %q for field %q, only types "string", "*string", []string and []*string supported as sensitive`, fieldType.String(), fieldNameCamel)
+				fieldType.String() != "[]*string" && fieldType.String() != "map[string]string" && fieldType.String() != "map[string]*string" {
+				return nil, nil, fmt.Errorf(`got type %q for field %q, only types "string", "*string", []string, []*string, "map[string]string" and "map[string]*string" supported as sensitive`, fieldType.String(), fieldNameCamel)
 			}
 			// Replace a parameter field with secretKeyRef if it is sensitive.
 			// If it is an observation field, it will be dropped.
@@ -170,9 +170,12 @@ func (g *Builder) buildResource(res *schema.Resource, cfg *config.Resource, tfPa
 				fieldType = typeSecretKeySelector
 			case "[]string", "[]*string":
 				fieldType = types.NewSlice(typeSecretKeySelector)
+			case "map[string]string", "map[string]*string":
+				fieldType = types.NewMap(types.Universe.Lookup("string").Type(), typeSecretKeySelector)
 			}
 			jsonTag = name.NewFromCamel(fieldNameCamel).LowerCamelComputed
-			if sch.Optional {
+			// Maps and slices are already pointers, so we don't need to wrap them even if they are optional.
+			if sch.Optional && sch.Type != schema.TypeMap && sch.Type != schema.TypeList {
 				fieldType = types.NewPointer(fieldType)
 				jsonTag += ",omitempty"
 			}

--- a/pkg/types/builder_test.go
+++ b/pkg/types/builder_test.go
@@ -327,7 +327,7 @@ func TestBuild(t *testing.T) {
 				},
 			},
 			want: want{
-				err: errors.Wrapf(fmt.Errorf(`got type %q for field %q, only types "string", "*string", []string and []*string supported as sensitive`, "*float64", "Key1"), "cannot build the Types"),
+				err: errors.Wrapf(fmt.Errorf(`got type %q for field %q, only types "string", "*string", []string, []*string, "map[string]string" and "map[string]*string" supported as sensitive`, "*float64", "Key1"), "cannot build the Types"),
 			},
 		},
 		"References": {

